src/main/java/com/devonfw/quarkus/productmanagement/domain/repo/ProductFragmentImpl.java

```diff
@@ -28,87 +24,37 @@
   EntityManager em;
 
   @Override
-  public Page<ProductEntity> findAllCriteriaApi(ProductSearchCriteriaDto dto) {
-
-    CriteriaQuery<ProductEntity> cq = this.em.getCriteriaBuilder().createQuery(ProductEntity.class);
-    Root<ProductEntity> root = cq.from(ProductEntity.class);
-    List<Predicate> predicates = new ArrayList<>();
-    CriteriaBuilder cb = this.em.getCriteriaBuilder();
-    if (dto.getTitle() != null && !dto.getTitle().isEmpty()) {
-      predicates.add(cb.like(root.get(ProductEntity_.TITLE), dto.getTitle()));
-    }
-
-    if (dto.getPrice() != null) {
-      predicates.add(cb.gt(root.get(ProductEntity_.PRICE), dto.getPrice()));
-    } else if (dto.getPriceMin() != null || dto.getPriceMax() != null) {
-      if (dto.getPriceMin() != null) {
-        predicates.add(cb.gt(root.get(ProductEntity_.PRICE), dto.getPriceMin()));
-      }
-      if (dto.getPriceMax() != null) {
-        predicates.add(cb.lt(root.get(ProductEntity_.PRICE), dto.getPriceMax()));
-      }
-    }
-
-    if (!predicates.isEmpty())
-
-    {
-      cq.where(predicates.toArray(new Predicate[0]));
-    }
-
-    // Order by title
-    cq.orderBy(cb.desc(root.get(ProductEntity_.TITLE)));
+  public Page<ProductEntity> findByCriteria(ProductSearchCriteriaDto searchCriteria) {
 
-    TypedQuery<ProductEntity> products = this.em.createQuery(cq).setFirstResult(dto.getPageNumber() * dto.getPageSize())
-        .setMaxResults(dto.getPageSize());
-    return new PageImpl<ProductEntity>(products.getResultList(), PageRequest.of(dto.getPageNumber(), dto.getPageSize()),
-        products.getResultList().size());
-  }
+    QProductEntity product = QProductEntity.productEntity;
 
-  @Override
-  public Page<ProductEntity> findAllQueryDsl(ProductSearchCriteriaDto dto) {
+    Predicate[] predicates = new Predicate[3];
```

Comment:
This is fragile solution for future. In case some developer adds a new criteria, he might forget to increase the size of the array. Additionally the array will most of the time have many items null(querydsl skips them)
Simply using `List` for collecting predicates and then at the end call `cq.where(predicates.toArray(Predicate[]::new));` is cleaner

Reply:
`cq.where(predicates.toArray(new Predicate[predicates.size()]))` is better as it prevents creating an empty array first that is only used to pass the actually array component type and then is replaced by a new array only feeding the GC.